html/semantics/forms/form-submission-0/jsurl-navigation-then-form-submit.html is failing in WebKit

cdumez opened the pull request:
#### 7de67f9a0ccc4fc8ce195e32f61e7873ad81345e
<pre>
html/semantics/forms/form-submission-0/jsurl-navigation-then-form-submit.html is failing in WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=243736">https://bugs.webkit.org/show_bug.cgi?id=243736</a>

Reviewed by Geoffrey Garen and Darin Adler.

The test was failing because the load event would fire prematurely for an iframe whose src
attribute is a JavaScript URL, before this JS URL has been executed. The reason for this is
that FrameLoader::SubframeLoader::requestFrame() would have a very special handling of JS
URLS which would call loadOrRedirectSubframe() with &quot;about:blank&quot; and then schedule a
navigation to the JS URL. The issue was that loading &quot;about:blank&quot; would trigger the load
event, since the iframe probably completed its first load.

To address the issue, we now drop the special handling for JS URLs inside requestFrame()
and instead go through the regular FrameLoader loading logic, even if the URL is a JS
URL. In PolicyChecker, we bypass doing the policy check with the client when the URL
is a JS URL, to maintain the previous behavior but we call the completion handler
asynchronously so that the JS URL is executed asynchronously, as is expected.

* LayoutTests/fast/loader/nested-document-handling-expected.txt:
* LayoutTests/fast/loader/nested-document-handling.html:
* LayoutTests/fast/repaint/repaint-during-scroll-with-zoom.html:
Update a couple of tests due to our behavior change. I have verified that those tests
were not passing in Chrome or Firefox before my changes. Our behavior on those tests
is now aligned with other major browsers.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/form-submission-0/jsurl-navigation-then-form-submit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/execution-timing/109-expected.txt:
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::executeJavaScriptURL):
* Source/WebCore/bindings/js/ScriptController.h:
(WebCore::ScriptController::executeJavaScriptURL):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::responseReceived):
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::shouldExecuteAsynchronouslyIfJavaScript const):
(WebCore::FrameLoadRequest::setShouldExecuteAsynchronouslyIfJavaScript):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::shouldExecuteJavaScriptURLSynchronously):
(WebCore::FrameLoader::loadURLIntoChildFrame):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadWithNavigationAction):
(WebCore::FrameLoader::executeJavaScriptURL):
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/FrameLoaderTypes.h:
(WebCore::PolicyCheckIdentifier::operator== const): Deleted.
(WebCore::PolicyCheckIdentifier::PolicyCheckIdentifier): Deleted.
(WebCore::PolicyCheckIdentifier::encode const): Deleted.
(WebCore::PolicyCheckIdentifier::decode): Deleted.
* Source/WebCore/loader/NavigationAction.h:
(WebCore::NavigationAction::shouldExecuteAsynchronouslyIfJavaScript const):
(WebCore::NavigationAction::setShouldExecuteAsynchronouslyIfJavaScript):
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::checkNavigationPolicy):
(WebCore::FrameLoader::PolicyChecker::checkNewWindowPolicy):
(WebCore::FrameLoader::PolicyChecker::stopCheck):
(WebCore::PolicyCheckIdentifier::create): Deleted.
(WebCore::PolicyCheckIdentifier::isValidFor): Deleted.
* Source/WebCore/loader/PolicyChecker.h:
* Source/WebCore/loader/SubframeLoader.cpp:
(WebCore::FrameLoader::SubframeLoader::requestFrame):
(WebCore::FrameLoader::SubframeLoader::loadOrRedirectSubframe):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):

Canonical link: <a href="https://commits.webkit.org/253350@main">https://commits.webkit.org/253350@main</a>
</pre>

geoffreygaren reviewed (approved):
r=me

This is a big change! But I think it's the right approach.

darinadler commented on Source/WebCore/loader/PolicyChecker.cpp:
```diff
@@ -76,19 +77,11 @@ static bool isAllowedByContentSecurityPolicy(const URL& url, const Element* owne
-    return m_policyCheck == expectedIdentifier.m_policyCheck;
+    // FIXME: Some sites rely on the javascript:'' loading synchronously, which is why we have this special case.
+    // Blink has the same workaround (https://bugs.chromium.org/p/chromium/issues/detail?id=923585).
+    return url == "javascript:''"_s || url == "javascript:\"\""_s;
```
Should this ignore ASCII case?

cdumez replied:
I am merely moving the code. This is a quirk and Blink seems to be doing a case sensitive check like us:
https://chromium-review.googlesource.com/c/chromium/src/+/1435821/

darinadler commented on Source/WebCore/loader/PolicyChecker.cpp:
```diff
@@ -76,19 +77,11 @@ static bool isAllowedByContentSecurityPolicy(const URL& url, const Element* owne
-    RELEASE_ASSERT_WITH_MESSAGE(m_process == expectedIdentifier.m_process, "Received a policy check response for a wrong process");
-    RELEASE_ASSERT_WITH_MESSAGE(m_policyCheck <= expectedIdentifier.m_policyCheck, "Received a policy check response from the future");
-    return m_policyCheck == expectedIdentifier.m_policyCheck;
+    // FIXME: Some sites rely on the javascript:'' loading synchronously, which is why we have this special case.
```
Is this tested by WPT?

cdumez replied:
I am not sure about WPT test coverage, we do have layout test coverage though. This is also not new behavior in this patch, just moving it.

darinadler commented on Source/WebCore/loader/PolicyChecker.cpp:
```diff
@@ -241,6 +234,26 @@ void FrameLoader::PolicyChecker::checkNavigationPolicy(ResourceRequest&& request
 
+    // Historically, we haven't asked the client for JS URL but we still want to execute them asynchronously.
+    if (requestIsJavaScriptURL) {
+        auto* document = m_frame.document();
```
RefPtr?

cdumez replied:
Ok.

darinadler commented on Source/WebCore/loader/SubframeLoader.cpp:
```diff
@@ -74,44 +74,17 @@ void FrameLoader::SubframeLoader::clear()
-            frame->script().executeJavaScriptURL(scriptURL);
-        else
-            frame->navigationScheduler().scheduleLocationChange(ownerElement.document(), ownerElement.document().securityOrigin(), scriptURL, m_frame.loader().outgoingReferrer(), lockHistory, lockBackForwardList, stopDelayingLoadEvent.release());
+    if (url.protocolIsJavaScript() && (!ownerElement.document().contentSecurityPolicy()->allowJavaScriptURLs(aboutBlankURL().string(), { }, url.string(), &ownerElement) || !ownerElement.canLoadScriptURL(url))) {
```
Normally we don’t use braces for if statements with single line bodies in WebKit coding style. Although this expression is so long! Almost makes me wish for a helper function.

cdumez replied:
Indeed, fixing.

cdumez committed:
html/semantics/forms/form-submission-0/jsurl-navigation-then-form-submit.html is failing in WebKit
https://bugs.webkit.org/show_bug.cgi?id=243736

Reviewed by Geoffrey Garen and Darin Adler.

The test was failing because the load event would fire prematurely for an iframe whose src
attribute is a JavaScript URL, before this JS URL has been executed. The reason for this is
that FrameLoader::SubframeLoader::requestFrame() would have a very special handling of JS
URLS which would call loadOrRedirectSubframe() with "about:blank" and then schedule a
navigation to the JS URL. The issue was that loading "about:blank" would trigger the load
event, since the iframe probably completed its first load.

To address the issue, we now drop the special handling for JS URLs inside requestFrame()
and instead go through the regular FrameLoader loading logic, even if the URL is a JS
URL. In PolicyChecker, we bypass doing the policy check with the client when the URL
is a JS URL, to maintain the previous behavior but we call the completion handler
asynchronously so that the JS URL is executed asynchronously, as is expected.

* LayoutTests/fast/loader/nested-document-handling-expected.txt:
* LayoutTests/fast/loader/nested-document-handling.html:
* LayoutTests/fast/repaint/repaint-during-scroll-with-zoom.html:
Update a couple of tests due to our behavior change. I have verified that those tests
were not passing in Chrome or Firefox before my changes. Our behavior on those tests
is now aligned with other major browsers.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/form-submission-0/jsurl-navigation-then-form-submit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/execution-timing/109-expected.txt:
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::executeJavaScriptURL):
* Source/WebCore/bindings/js/ScriptController.h:
(WebCore::ScriptController::executeJavaScriptURL):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::responseReceived):
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::shouldExecuteAsynchronouslyIfJavaScript const):
(WebCore::FrameLoadRequest::setShouldExecuteAsynchronouslyIfJavaScript):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::shouldExecuteJavaScriptURLSynchronously):
(WebCore::FrameLoader::loadURLIntoChildFrame):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadWithNavigationAction):
(WebCore::FrameLoader::executeJavaScriptURL):
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/FrameLoaderTypes.h:
(WebCore::PolicyCheckIdentifier::operator== const): Deleted.
(WebCore::PolicyCheckIdentifier::PolicyCheckIdentifier): Deleted.
(WebCore::PolicyCheckIdentifier::encode const): Deleted.
(WebCore::PolicyCheckIdentifier::decode): Deleted.
* Source/WebCore/loader/NavigationAction.h:
(WebCore::NavigationAction::shouldExecuteAsynchronouslyIfJavaScript const):
(WebCore::NavigationAction::setShouldExecuteAsynchronouslyIfJavaScript):
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::checkNavigationPolicy):
(WebCore::FrameLoader::PolicyChecker::checkNewWindowPolicy):
(WebCore::FrameLoader::PolicyChecker::stopCheck):
(WebCore::PolicyCheckIdentifier::create): Deleted.
(WebCore::PolicyCheckIdentifier::isValidFor): Deleted.
* Source/WebCore/loader/PolicyChecker.h:
* Source/WebCore/loader/SubframeLoader.cpp:
(WebCore::FrameLoader::SubframeLoader::requestFrame):
(WebCore::FrameLoader::SubframeLoader::loadOrRedirectSubframe):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):

Canonical link: https://commits.webkit.org/253350@main
